### PR TITLE
Account Management System

### PR DESCRIPTION
**Pull request by lantingchiang**
This PR adds an account management system for users to verify methods of contact, add a preferred name, and add major/school information.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -28,6 +28,10 @@ uwsgi = "*"
 uritemplate = "*"
 shortener = "*"
 django-runtime-options = "*"
+django-phonenumber-field = "*"
+phonenumbers = "*"
```
Comment (ArmaanT) [inline Pipfile:32]: You should be able to install `phonenumbers` through `django-phonenumber-field`

File: accounts/admin.py [hunk 2]
```diff
@@ -32,18 +32,28 @@ class UserAdmin(admin.ModelAdmin):
...
         (
             ("Permissions"),
-            {"fields": ("is_active", "is_staff", "is_superuser", "groups", "user_permissions")},
+            {"fields": ("is_active", "is_staff", "is_superuser", "groups", "user_permissions",)},
         ),
         (("Important dates"), {"fields": ("last_login", "date_joined")}),
     )
 
 
+class PhoneNumberAdmin(admin.ModelAdmin):
```
Comment (ArmaanT) [inline accounts/admin.py:44]: If we're not adding custom functionality to them we don't need to make the `PhoneNumberAdmin` and `EmailAdmin`

File: accounts/admin.py [hunk 3]
```diff
@@ -32,18 +32,28 @@ class UserAdmin(admin.ModelAdmin):
     ordering = ("username",)
     fieldsets = (
         (None, {"fields": ("username", "pennid")}),
-        (("Personal info"), {"fields": ("first_name", "last_name", "email")}),
+        (("Personal info"), {"fields": ("preferred_name", "first_name", "last_name", "email")},),
```
Comment (ArmaanT) [inline accounts/admin.py:35]: We should probably keep first name in a read-only field just so we can refer to it if needed

File: accounts/migrations/0003_auto_20200421_2056.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# Generated by Django 3.0.3 on 2020-04-22 00:56
```
Comment (ArmaanT) [inline accounts/migrations/0003_auto_20200421_2056.py:1]: Squash all these migrations before merging

File: accounts/models.py [hunk 5]
```diff
@@ -3,19 +3,56 @@
...
+    def get_preferred_name(self):
+        if self.preferred_name != "":
+            return self.preferred_name
+        else:
+            return self.first_name
 
 
 class Student(models.Model):
     user = models.OneToOneField(get_user_model(), on_delete=models.DO_NOTHING)
     major = models.CharField(max_length=255, blank=True)
```
Comment (ArmaanT) [inline accounts/models.py:27]: We still need to convert major and school into separate models, right? @JayVish

File: accounts/models.py [hunk 6]
```diff
@@ -3,19 +3,56 @@
...
     school = models.CharField(max_length=255, blank=True)
+    graduation_year = models.IntegerField(null=True)
```
Comment (ArmaanT) [inline accounts/models.py:29]: Should this be blank instead of null? I never really got a good idea of that

File: accounts/serializers.py [hunk 7]
```diff
@@ -25,12 +30,38 @@ class Meta:
...
+            "email",
+            "groups",
+            "product_permission",
+            "user_permissions",
+        )
+
+    # Users are pulled from Penn DB, so come with no preferred
+    # name. Thus, this logic only needs to happen on update.
+    def update(self, instance, validated_data):
+        if validated_data["get_preferred_name"]:
```
Comment (ArmaanT) [inline accounts/serializers.py:46]: this should be `if "get_preferred_name" in validated_data`

Comment (ArmaanT) [reply accounts/serializers.py:46]: Also, I'm not too happy with how the logic is structured here. I know we came up with it together, but it feels like there should be a slightly cleaner way of doing things.

File: accounts/templates/emails/base.html [hunk 8]
```diff
@@ -0,0 +1,78 @@
+<head>
```
Comment (ArmaanT) [inline accounts/templates/emails/base.html:1]: Note to self: we need to customize these for platform

File: accounts/views.py [hunk 9]
```diff
@@ -156,6 +162,109 @@ def get_queryset(self):
...
+    You must specify all of the fields or use a patch request.
+
+    patch:
+    Update information about the logged in user.
+    Only updates fields that are passed to the server.
+    """
+
+    serializer_class = UserSerializer
+    permission_classes = [IsAuthenticated]
+    # lookup_field = "pennid"
```
Comment (ArmaanT) [inline accounts/views.py:181]: Remove comment

File: accounts/views.py [hunk 10]
```diff
@@ -156,6 +162,109 @@ def get_queryset(self):
...
+        return self.request.user.phone_numbers.all()
+
+    def destroy(self, request, *args, **kwargs):
+        is_primary = self.get_object().primary
+        self.get_object().delete()
+        if is_primary and len(self.get_queryset().filter(verified=True)) > 0:
+            next_number = self.get_queryset().filter(verified=True)[0]
+            next_number.primary = True
+            next_number.save()
+        return HttpResponse("Number successfully deleted", status=200)
```
Comment (ArmaanT) [inline accounts/views.py:223]: All the `HttpResponse`s should be JsonResponses instead

File: accounts/views.py [hunk 11]
```diff
@@ -156,6 +162,109 @@ def get_queryset(self):
...
+    permission_classes = [IsAuthenticated]
+
+    def get_queryset(self):
+        return self.request.user.phone_numbers.all()
+
+    def destroy(self, request, *args, **kwargs):
+        is_primary = self.get_object().primary
+        self.get_object().delete()
+        if is_primary and len(self.get_queryset().filter(verified=True)) > 0:
+            next_number = self.get_queryset().filter(verified=True)[0]
```
Comment (ArmaanT) [inline accounts/views.py:220]: You can simplify these lines with something like
```python
next = self.get_queryset().filter(verified=True).first()
if next is not None:
    ...
```

File: accounts/views.py [hunk 12]
```diff
@@ -156,6 +162,109 @@ def get_queryset(self):
...
+
+    serializer_class = EmailSerializer
+    permission_classes = [IsAuthenticated]
+
+    def get_queryset(self):
+        return self.request.user.emails.all()
+
+    def destroy(self, request, *args, **kwargs):
+        if len(self.get_queryset().filter(verified=True)) < 2:
+            return HttpResponse("You can't delete the only verified email", status=405)
```
Comment (ArmaanT) [inline accounts/views.py:257]: Same Http -> Json comment

File: accounts/views.py [hunk 13]
```diff
@@ -156,6 +162,109 @@ def get_queryset(self):
...
+
+        is_primary = self.get_object().primary
+        self.get_object().delete()
+        if is_primary and len(self.get_queryset().filter(verified=True)) > 0:
+            next_email = self.get_queryset().filter(verified=True)[0]
+            next_email.primary = True
+            next_email.save()
+        return HttpResponse("Email successfully deleted", status=200)
```
Comment (ArmaanT) [inline accounts/views.py:265]: Same Http -> Json comment

File: accounts/views.py [hunk 14]
```diff
@@ -156,6 +162,109 @@ def get_queryset(self):
...
+    def get_queryset(self):
+        return self.request.user.emails.all()
+
+    def destroy(self, request, *args, **kwargs):
+        if len(self.get_queryset().filter(verified=True)) < 2:
+            return HttpResponse("You can't delete the only verified email", status=405)
+
+        is_primary = self.get_object().primary
+        self.get_object().delete()
+        if is_primary and len(self.get_queryset().filter(verified=True)) > 0:
```
Comment (ArmaanT) [inline accounts/views.py:261]: same logic simplification suggestion as above

File: accounts/views.py [hunk 15]
```diff
@@ -156,6 +162,109 @@ def get_queryset(self):
...
+    """
+
+    serializer_class = EmailSerializer
+    permission_classes = [IsAuthenticated]
+
+    def get_queryset(self):
+        return self.request.user.emails.all()
+
+    def destroy(self, request, *args, **kwargs):
+        if len(self.get_queryset().filter(verified=True)) < 2:
```
Comment (ArmaanT) [inline accounts/views.py:256]: In general, you shouldn't call len() on a queryset. You should use qs.count() instead because that makes a more efficient SQL query.

File: tests/accounts/test_serializers.py [hunk 16]
```diff
@@ -25,7 +38,22 @@ def setUp(self):
         self.user.student.school = "School"
         self.serializer = StudentSerializer(self.user.student)
 
-    def test_str(self):
+        self.user_preferred_name = get_user_model().objects.create_user(
```
Comment (ArmaanT) [inline tests/accounts/test_serializers.py:41]: Note to self: We need to delete the org and services apps, and then refactor the StudentSerializer to just list out major, school and year.

File: tests/accounts/test_views.py [hunk 17]
```diff
@@ -191,3 +198,147 @@ def test_first_name(self):
...
+        self.client.force_authenticate(user=self.user)
+        response = self.client.get(reverse("accounts:me-phonenumber-list"))
+        self.assertEqual(
+            json.loads(response.content),
+            [self.serializer1.data, self.serializer2.data, self.serializer3.data],
+        )
+
+    def test_destroy_nonprimary(self):
+        self.client.force_authenticate(user=self.user)
+        response = self.client.delete(reverse("accounts:me-phonenumber-detail", args=[1]))
```
Comment (ArmaanT) [inline tests/accounts/test_views.py:255]: Don't hardcode numbers here, do something like `self.number1.id`

Comment (ArmaanT) [reply tests/accounts/test_views.py:255]: And same for all the other places like this

**Review by ArmaanT — CHANGES_REQUESTED** [review]
We're almost there! Mostly have smaller comments on a few things

File: accounts/admin.py (hunk 3 above)
Comment (lantingchiang) [reply accounts/admin.py:35]: Moved last_name to read_only as well

File: accounts/models.py (hunk 6 above)
Comment (lantingchiang) [reply accounts/models.py:29]: null=True allows null values in the previously existing rows in the db that don't have a value for graduation_year to be null I think (?), so if we don't do null=True we'll want to set default=something. blank=True makes this field optional?

File: accounts/models.py (hunk 5 above)
Comment (lantingchiang) [reply accounts/models.py:27]: Should I remove major & school from the Student model then? @ArmaanT

Comment (JayVish) [reply accounts/models.py:27]: @ArmaanT Sorry, I hadn't seen this before. Yeah, I've converted majors into a separate major model on the add-major-school branch. I am working on doing the same for the school model as well.

Comment (lantingchiang) [reply accounts/models.py:27]: So if we map from the major & school model to the student model via one-to-one field, I can just remove major & school fields from the student model side?

Comment (ArmaanT) [reply accounts/models.py]: I think what we should do is wait for @JayVish to finish up the school/major features on his branch, then merge those into this one and then merge this into master

File: accounts/admin.py [hunk 18]
```diff
@@ -24,26 +24,41 @@ def last_name(self, obj):
...
+    readonly_fields = ("username", "first_name", "last_name", "pennid", "last_login", "date_joined")
     search_fields = ("username", "first_name", "last_name")
     list_display = ("username", "first_name", "last_name", "is_staff")
     list_filter = ("is_staff", "is_superuser", "is_active", "groups")
     filter_horizontal = ("groups", "user_permissions")
     ordering = ("username",)
     fieldsets = (
         (None, {"fields": ("username", "pennid")}),
-        (("Personal info"), {"fields": ("first_name", "last_name", "email")}),
+        (("Personal info"), {"fields": ("preferred_name", "email")},),
```
Comment (ArmaanT) [inline accounts/admin.py:35]: I think we should probably undo the changes to readonly_fields and add first/last name back to the personal info section just for the sake of an easier experience in local dev

**Review by ArmaanT — CHANGES_REQUESTED** [review]
Got a few minor comments on this PR. @JayVish do you want to tackle them? If not I can make them later this week.

**Comment by ArmaanT** [comment]
We also need to make a couple of modifications so that whenever a user is created, a student is also created (post save hook would probably be best here) and remove code from the backend that used to do that

**Comment by ArmaanT** [comment]
And we need to somehow automatically create an email object on first login with the email provided by shib and sync that info back to the user account

**Comment by JayVish** [comment]
> Got a few minor comments on this PR. @JayVish do you want to tackle them? If not I can make them later this week.

Yeah I'll look into these soon. I'll try to fix what I can and push it.

File: accounts/models.py [hunk 19]
```diff
@@ -2,20 +2,103 @@
...
+    primary = models.BooleanField(default=False)
+    verification_code = models.CharField(max_length=6, blank=True, null=True)
+    verification_timestamp = models.DateTimeField(blank=True, null=True)
+    verified = models.BooleanField(default=False)
+
+    def __str__(self):
+        return f"{self.user} - {self.value}"
+
+
+class PhoneNumberModel(models.Model):
```
[the remark this answers is not in the document]
Comment (JayVish) [reply accounts/models.py:93]: fixed

File: Pipfile [hunk 20]
```diff
@@ -30,6 +30,12 @@ uwsgi = "*"
 uritemplate = "*"
 shortener = "*"
 django-runtime-options = "*"
+django-phonenumber-field = {extras = ["phonenumbers"],version = "*"}
+django-email-tools = "*"
+twilio = "*"
+phonenumbers = "*"
```
[the remark this answers is not in the document]
Comment (JayVish) [reply Pipfile:36]: fixed

File: accounts/models.py [hunk 21]
```diff
@@ -2,20 +2,103 @@
...
+    degree_type = models.CharField(
+        max_length=20, choices=DEGREE_CHOICES, default=DEGREE_PROFESSIONAL
+    )
+
+    def __str__(self):
+        return self.name
 
 
 class User(AbstractUser):
+
```
[the remark this answers is not in the document]
Comment (JayVish) [reply accounts/models.py:49]: fixed

File: tests/accounts/test_commands.py [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+
+from accounts.models import Major, School
+
+
+class UpdateMajorsTestCase(TestCase):
+    def test_update_academics(self):
+        call_command("update_academics")
+
+        self.assertTrue(Major.objects.all().count() != 0)
+        self.assertTrue(School.objects.all().count() != 0)
```
[the remark this answers is not in the document]
Comment (JayVish) [reply tests/accounts/test_commands.py:12]: fixed

**Comment by ArmaanT** [comment]
TODO: figure out email templates

File: accounts/views.py [hunk 23]
```diff
@@ -156,6 +166,108 @@ def get_queryset(self):
...
+
+    serializer_class = EmailSerializer
+    permission_classes = [IsAuthenticated]
+
+    def get_queryset(self):
+        return self.request.user.emails.all()
+
+    def destroy(self, request, *args, **kwargs):
+        if self.get_queryset().filter(verified=True).count() < 2:
+            return Response({"detail": "You can't delete the only verified email"}, status=405)
```
Comment (eyingxuan) [inline accounts/views.py:260]: Should this check if what we are destroying is a verified email? This does not allow an unverified email to be deleted when there is only one verified email.

Comment (ArmaanT) [reply accounts/views.py:260]: That's a good catch. I'll push a fix in a sec

**Comment by ArmaanT** [comment]
TODO: after final review, make CI changes and refactor repo to have backend and frontend folders

**Review by ArmaanT — APPROVED** [review]
Here we go
